### DISCUSSION TITLE
4  Unconfigured Buttons in Home App

### DISCUSSION THREAD

**Discussion by almonde** [thoukydides/homebridge-homeconnect#177] [Q&A] [answered]
Firstly, thanks for creating and maintaining this HomeBridge plugin. Its must be extra annoying trying to find ways to shoehorn appliances that are not natively supported appliance types in using switches, contact sensors etc.  I think my question maybe related to this.

After installing the dishwasher I have 2 grouped tiles in HomeKit, one is a set of switches for programs, power state, and door state. All named and seems to work well.  I also have a separate tile which contacts '4 Unconfigured Buttons'. I believe these are the HomeKit stateless buttons.  My guess is they are meant to be there to show rinse aid state etc, but since they are not configured I'm not sure how I'm meant to designate the button to the state.

Thanks again.

**Comment by thoukydides** [comment, marked as answer]
This gets asked a lot, e.g. in issue #31.

The “unconfigured buttons” are [Stateless Programmable Switch services](https://github.com/thoukydides/homebridge-homeconnect/wiki/HomeKit-Mapping#stateless-programmable-switch-services-for-events) used to indicate Home Connect events that can be used to trigger HomeKit automations. They are stateless because the Home Connect API only reports them as events when they happen - it is not possible to read a persistent state.

The [README](https://github.com/thoukydides/homebridge-homeconnect#appliance-support) summarises what these are used for with different appliance types:
* CoffeeMaker: Automation triggers for bean container empty, water tank empty, and drip tray full.
* CookProcessor: Automation triggers for finished and aborted.
* Hob: Automation triggers for finished, alarm clock finished, and preheat finished.
* Hood: Automation triggers for finished, grease filter nearly saturated, and grease filter saturated.
* Oven: Automation triggers for finished, aborted, alarm clock finished, fast preheat finished, and regular preheat finished.
* CleaningRobot: Automation triggers for finished, aborted, dust box full, stuck, and lost.
* Dishwasher: Automation triggers for finished, aborted, salt supply low, and rinse aid supply low.
* Dryer: Automation triggers for finished and aborted.
* Washer / WasherDryer: Automation triggers for finished, aborted, i-Dos 1 fill level low, and i-Dos 2 fill level low.
* Freezer / FridgeFreezer / Refrigerator: Automation triggers for freezer/refrigerator door and freezer temperature alarms.

Unfortunately the Apple Home app only labels these buttons numerically so it isn’t obvious what each button indicates - you need to know the mapping. In the case of a dishwasher this can be found in [src/appliance_cleaning.js](https://github.com/thoukydides/homebridge-homeconnect/blob/c137ada629b486b8821a81c0d54483cb59a745a5/src/appliance_cleaning.js#L49):
```Javascript
      this.mixin(HasEvents, {
            'BSH.Common.Event.ProgramFinished':                 'Program Finished',
            'BSH.Common.Event.ProgramAborted':                  'Program Aborted',
            'Dishcare.Dishwasher.Event.SaltNearlyEmpty':        'Salt Low',
            'Dishcare.Dishwasher.Event.RinseAidNearlyEmpty':    'Rinse Aid Low'
        });
```

so the four buttons are:
1. Program Finished
2. Program Aborted
3. Salt Low
4. Rinse Aid Low

Some [third-party HomeKit apps](https://github.com/thoukydides/homebridge-homeconnect/wiki/HomeKit-Apps) are more helpful and show the actual functionality. For example, this is in [The Eve App](https://www.evehome.com/en/eve-app):
![IMG_0622](https://github.com/thoukydides/homebridge-homeconnect/assets/4775242/b4b754fa-1e18-41c3-a47b-ba9cabb6f3cc)

Hopefully the recent release of [Matter 1.2](https://csa-iot.org/newsroom/matter-1-2-arrives-with-nine-new-device-types-improvements-across-the-board/) will result in Apple adding more appropriate characteristics and services to HomeKit…

**Comment by thoukydides** [comment]
[v0.32.0](https://github.com/thoukydides/homebridge-homeconnect/releases/tag/v0.32.0) provides per-appliance configuration options to remove the `Stateless Programmable Switch` services for appliance events (as well as selectively enabling the `Switch` services for appliance mode settings and the `Door` services for the appliance door status).
